src/librustc_middle/mir/interpret/pointer.rs

```diff
@@ -48,26 +54,31 @@ pub trait PointerArithmetic: HasDataLayout {
 
     #[inline]
     fn overflowing_signed_offset(&self, val: u64, i: i64) -> (u64, bool) {
-        if i < 0 {
-            // Trickery to ensure that `i64::MIN` works fine: compute `n = -i`.
-            // This formula only works for true negative values; it overflows for zero!
-            let n = u64::MAX - (i as u64) + 1;
+        let n = uabs(i);
+        if i >= 0 {
+            self.overflowing_offset(val, n)
+        } else {
             let res = val.overflowing_sub(n);
             self.truncate_to_ptr(res)
-        } else {
-            // `i >= 0`, so the cast is safe.
-            self.overflowing_offset(val, i as u64)
         }
     }
 
     #[inline]
     fn offset<'tcx>(&self, val: u64, i: u64) -> InterpResult<'tcx, u64> {
+        // The unsigned offset must fit in a usize.
+        if i > self.machine_usize_max() {
+            throw_ub!(PointerArithOverflow)
+        }
```

Comment:
This doesn't seem right -- `overflowing_offset` will still return `false` if such an overflow occurs, right?

I think the check should probably happen in the `overflowing_*` methods instead.

Reply:
You're right, I'll add the checks to `overflowing_offset` and `overflowing_signed_offset`.

Reply:
Done